src/Periphery/ReceiverChainflip.sol

```diff
@@ -0,0 +1,154 @@
+    constructor(
+        address _owner,
+        address _executor,
+        address _chainflipVault
+    ) WithdrawablePeriphery(_owner) {
+        executor = IExecutor(_executor);
+        chainflipVault = _chainflipVault;
+    }
+
+    /// External Methods ///
+
+    /// @notice Receiver function for Chainflip cross-chain messages
+    /// @dev This function can only be called by the Chainflip Vault on this network
+    /// @dev First param (unused): The source chain according to Chainflip's nomenclature
+    /// @dev Second param (unused): The source address on the source chain
+    /// @param message The message sent from the source chain
+    /// @param token The address of the token received
+    /// @param amount The amount of tokens received
+    function cfReceive(
+        uint32, // srcChain
+        bytes calldata, // srcAddress
+        bytes calldata message,
+        address token,
+        uint256 amount
+    ) external payable onlyChainflipVault {
+        // decode payload
+        (
+            bytes32 transactionId,
+            LibSwap.SwapData[] memory swapData,
+            address receiver
+        ) = abi.decode(message, (bytes32, LibSwap.SwapData[], address));
+
+        // execute swap(s)
+        _swapAndCompleteBridgeTokens(
+            transactionId,
+            swapData,
+            token,
+            payable(receiver),
+            amount
+        );
+    }
+
+    /// Private Methods ///
+
+    /// @notice Performs a swap before completing a cross-chain transaction
+    /// @param _transactionId the transaction id associated with the operation
+    /// @param _swapData array of data needed for swaps
+    /// @param assetId address of the token received from the source chain
+    /// @param receiver address that will receive tokens in the end
+    /// @param amount amount of token
+    /// @notice Performs a swap before completing a cross-chain transaction
+    /// @param _transactionId The transaction id associated with the operation
+    /// @param _swapData Array of data needed for swaps
+    /// @param assetId Address of the token received from the source chain
+    /// @param receiver Address that will receive tokens in the end
+    /// @param amount Amount of token to swap
+    /// @dev If the swap fails, the original bridged tokens are sent directly to the receiver
+    /// @dev Approvals are reset to 0 after the operation completes
+    function _swapAndCompleteBridgeTokens(
+        bytes32 _transactionId,
+        LibSwap.SwapData[] memory _swapData,
+        address assetId,
+        address payable receiver,
+        uint256 amount
+    ) private {
+        // Convert Chainflip's native token address to LibAsset.NATIVE_ASSETID
+        address actualAssetId = assetId == CHAINFLIP_NATIVE_ADDRESS
+            ? LibAsset.NATIVE_ASSETID
+            : assetId;
+
+        // Don't need approval for native token
+        if (actualAssetId != LibAsset.NATIVE_ASSETID) {
```

Comment:
I think you could (gas-)optimize this code:

```solidity
address actualAssetId;
if(assetId == CHAINFLIP_NATIVE_ADDRESS) {
 actualAssetId=LibAsset.NATIVE_ASSETID;
} else {
 actualAssetId=assetId;
 actualAssetId.safeApproveWithRetry(address(executor), amount);
}
```
(saves one comparison)

you could go even one step further and include the specific try/catch statement with their custom fallback handling depending on the token type as it's better for our gas consumption to deploy more code but have less checks/comparisons during runtime.